### ISSUE TITLE
Big Rescaling + O(0.001) IC + Complicated-Shape Domain

### ISSUE THREAD

**Issue by Ryszard2** [lululxvi/deepxde#345]
Good morning Dr. @lululxvi,

I got stuck in a problem I already asked about some time ago. It's a Riemann Problem of the Shallow Water Equations where the domain is complicated and big:

- Lmax = 16,500 m
- T = 2,000 s

That's the original IC for the Water Height:

![IC Original](https://user-images.githubusercontent.com/79721174/126759417-2ba863e0-a87f-446d-8ad8-3e9c7b81a6c2.png)

#

I massively rescaled the problem in such a way to keep the **equations unchanged**, I decided to do it this way because otherwise some coefficients very very far from unitary value emerge in the equations.
Now:

- Lmax = 1 m
- T = 15.57 s
- IC unfortunately becomes O(0.001)

![IC Scaled](https://user-images.githubusercontent.com/79721174/126760030-38f7c199-12ca-4e10-a35c-99e1f49bdf76.png)

#

The problems are that:

- I can't get the training to convergence in no way at all, basically because the Loss for the IC for the height can't get down
- I can't hard-constrain the ICs nor the BCs because otherwise I get nans
- Trying and trying I kind of concluded that I need a lot of training points, something that can be done with the small-batch resampling, but still the IC for the height keeps the Loss pretty high
- The input data come all in the form of **matrices**, that seems definitely to be a problem for the learning of the IC

The code seems pretty correct to me, I can't figure out at all what I'm missing to get the job done.

#

Thank you
Riccardo

**Comment by lululxvi** [comment]
As you discovered, the first issue could be the training of IC. Could you try only with IC loss to see whether the IC can be trained well? You can set the PDE to be None or the weight of PDE is 0.

**Comment by Ryszard2** [comment]
Thank you @lululxvi 

Say I run a couple-hundred-thousands epochs to learn the one IC that bothers me (there are 3 ICs, but the other 2 are not a big deal), and I end up with a train loss like that:

`[0.00e+00, 0.00e+00, 0.00e+00, 4.19e-07, 0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00]`

as a result of this loss weights setup:

`model.compile('adam', lr=0.001, loss_weights=[0, 0, 0, 1e2, 0, 0, 0, 0])`

How do I proceed in the next training phase?
Am I supposed to set the remaining weights so that every component of the loss is O(1e-7) to match the loss I've achieved with the IC component?

By the way, I settled for a `[50]*4` NN, I can't afford a bigger NN :-(

#

Thank you
Riccardo

**Comment by Ryszard2** [comment]
Another question, @lululxvi 

is it reasonable to sample 1,000 training points, let the training go for like 20,000 epochs, then sample 1,000 new training points for other 20,000 epochs and going this way up to some hundred thousands epochs?

**Comment by lululxvi** [comment]
You can use the two-stage training in this paper: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007575 with the code https://github.com/alirezayazdani1/SBINNs.

Yes, resampling makes sense. Here is an example: https://github.com/lululxvi/deepxde/blob/master/examples/diffusion_1d_resample.py using dde.callbacks.PDEResidualResampler

**Comment by lululxvi** [comment]
The resampling period seems OK. You may increase the other three numbers by a factor of 10.

**Comment by Ryszard2** [comment]
Thank you @lululxvi

First of all, I totally scaled the IC like this:

![index](https://user-images.githubusercontent.com/79721174/131624761-fa294108-7bc2-4670-9a9e-1f7d795a49c0.png)

But still the learning of it is a challenge I never faced before. I did it so far without small-batch, only by setting to zero the non-BC loss weights. After 600k epochs, still some bothering non-physical negative water remains, stucked in the narrow coves near the border

![ic1](https://user-images.githubusercontent.com/79721174/131625697-f1d4cba9-69f2-4c9e-a801-4def07bf86c7.png)

Now I'm going all-in for other hundreds of thousands of epochs, working on the good-but-not-great model I already got to after 600k epochs, for IC-BC only with this setiing:

```python
num_domain   = 0,
num_boundary = 400,
num_initial  = 400,

resampler = dde.callbacks.PDEResidualResampler(period=100)
```

hoping for a faster learning of this absurd IC :-(
I prefer not to go on with the learning of the PDE outputs until the IC is properly learned.

**Comment by lululxvi** [comment]
To clarify, you currently only learn the IC and BC, right? Does the full gradient descent with 10,000 IC points work?

**Comment by Ryszard2** [comment]
The moment I got I could go big with small-batch I started over the training. I started (and still am) learning only one IC, I'm working like this:

```python
data = dde.data.TimePDE(
    geomtime, pde, BC,
    num_domain   = 0,
    num_boundary = 0,
    num_initial  = 12000,
    auxiliary_var_function = auxiliary_var_function)
```

```python
resampler = dde.callbacks.PDEResidualResampler(period=100)
```

```python
model.compile('adam', lr=0.0001, loss_weights=[0, 0, 0, 1, 0, 0, 0, 0])
```

The result after 660,000 epochs is honest but I can't say it's satisfying, since some spots with non-physical negative-height water still show up:

![1](https://user-images.githubusercontent.com/79721174/132137146-5299537e-3cf0-4673-8c69-cb50e1dddd87.png)

![2](https://user-images.githubusercontent.com/79721174/132137147-4c312d8e-f964-4744-ac61-43d9cb21f081.png)

![3](https://user-images.githubusercontent.com/79721174/132137149-2d0d99a2-2c2b-4450-a569-7a13cb59f8df.png)

The loss is:

```python
[0.00e+00, 0.00e+00, 0.00e+00, 1.40e-07, 0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00]
```

I may be naive, but after 660,000 epochs I would expect at least one IC to be learned, but this is not the case. I point out that I'm working on a `[50]*4` NN, I'm starting to think that maybe this architecture is pretty inadequate for the complexity of this very problem.

@lululxvi  would you suggest to get out of the comfort zone and embrace a monster `[128]*4` NN or (boy, no!) a `[256]*5` NN?

**Comment by Ryszard2** [comment]
> Does the full gradient descent with 10,000 IC points work?

It can bring to an honest result, still far from good.

I made a simple test: the share of points that goes to the complex non-homogeneous sector is less than 10% of the points generated, i.e. the 90% of the training points are placed in the easy homogeneous sector. This is a pretty big waste, so I anchor the points by myself with the `random_points` generator.

![ic](https://user-images.githubusercontent.com/79721174/132367764-f7d24878-b954-4f41-b4ac-9e8ffecbc2fe.png)

With such training points' number and distribution, with a large NN, the full gradient descent make totally sense, and I'm sure that the learning will be perfect.

**Comment by Ryszard2** [comment]
@lululxvi I made a simple test and I realized that the share of sampled points that goes to the complex non-homogeneous little sector is less than 10% of the points generated, i.e. the 90% of the training points are placed in the easy homogeneous big sector!

This is a pretty big waste, so I decided to anchor the points by myself with the `random_points` generator:

![ic](https://user-images.githubusercontent.com/79721174/132367764-f7d24878-b954-4f41-b4ac-9e8ffecbc2fe.png)

With such training points' number and distribution - with the `[256]*5` NN I'm finally using - the full gradient descent makes totally sense and I'm sure that the learning of the IC will be perfect.

Still, the spatio-temporal problem of PDE and BC remains.
The highly irregular distribution of IC forces me to choose the training points, I'll anchor the points by myself but I can't process a full gradient descent with that amount of points, so I'd ask you:

- say that I anchor a zillion points in the spatio-temporal domain, would it be possible for DeepXDE to split that array in small batches? I would like to:

  1. create the giant array of training point (I'd maybe shuffle its elements' order, afterwards)
  2. ask DeepXDE to take 1,000 points at the time, train 'em for 100 epochs, then take the next 1,000. So on up till the end of the array...

Is it possible? Honestly I can't see another way to tacke this problem.
I just point out that, after a massive rescaling, the Final Time is 1 second.

**Comment by lululxvi** [comment]
It is great to see that you made some good progress.

If the training points are sampled by yourself, then DeepXDE won't split them into batches. Currently DeepXDE only splits the points that sampled by DeepXDE. Here are a few thoughs:
- Actually I think you may not need a zillion points. Usually, a relatively dense and fixed points works well enough, and there is no need to use lots of batches of differenet points.
- If you still want to use batches, then you may have to modify the function:
https://github.com/lululxvi/deepxde/blob/6999a9d12a9218c86db45a0b036d0a8900a8610a/deepxde/data/pde.py#L188
to define how to resample each time by yourself.
Then you can use it together with `dde.callbacks.PDEResidualResampler`

**Comment by Ryszard2** [comment]
Indeed, @lululxvi, I was actually thinking about a kinda trade-off:

- I sample the residual points by myself - with the nonuniform distribution I desire - ONLY for the IC (and maybe the BC too, I'll think about it)
- I let DeepXDE the sampling/resampling of the domain points

In my idea, I would make sure to learn well the complicated IC. There wouldn't be strict need of my anchoring afterwards, since the hyperbolic nature of the problem would move the solution downstream, where the polygon widens and the training points are naturally more.

**Comment by Rdfing** [comment]
@Ryszard2 I am very interested in this project. Could you please share a link to your paper once you get the DeepXDE working on 2D SWE? Thanks, Haochen

**Comment by Ryszard2** [comment]
@Rdfing actually this is a test case for my MSc thesis, a considerably challenging case.

**Comment by Rdfing** [comment]
> @Rdfing actually this is a test case for my MSc thesis, a considerably challenging case.

Wow, that is impressive! I get SWE to work with PINNs for some simple 2D benchmark cases, but have not tried for this benchmark.

**Comment by lululxvi** [comment]
@FZUcipher See FAQ Q: I failed to train the network or get the right solution, e.g., large training loss, unbalanced losses.

**Comment by thevelvetunderground** [comment]
> Thank you @lululxvi
> 
> First of all, I totally scaled the IC like this:
> 
> ![index](https://user-images.githubusercontent.com/79721174/131624761-fa294108-7bc2-4670-9a9e-1f7d795a49c0.png)
> 
> But still the learning of it is a challenge I never faced before. I did it so far without small-batch, only by setting to zero the non-BC loss weights. After 600k epochs, still some bothersome non-physical negative water remains, stucked in the narrow coves near the border
> 
> ![ic1](https://user-images.githubusercontent.com/79721174/131625697-f1d4cba9-69f2-4c9e-a801-4def07bf86c7.png)
> 
> Now I'm going all-in for other hundreds of thousands of epochs, working on the good-but-not-great model I already got to after 600k epochs, for IC-BC only with this setting:
> 
> ```python
> num_domain   = 0,
> num_boundary = 400,
> num_initial  = 400,
> 
> resampler = dde.callbacks.PDEResidualResampler(period=100)
> ```
> 
> hoping for a faster learning of this absurd IC :-( I prefer not to go on with the learning of the PDE outputs until the IC is properly learned.
> 
> Edit: It didn't work out, this setting got the learning of the IC way worse :-(
> 
> Re-Edit: I noticed that setting the training like this:
> 
> ```python
> num_domain   = 0,
> num_boundary = 1000,
> num_initial  = 4000,
> 
> resampler = dde.callbacks.PDEResidualResampler(period=100)
> ```
> 
> works a little better. Not as good as a full gradient descent with 10,000 IC points, but seems like sampling a zillion IC points every 100 epochs, for this specific case, could bring to something good. I can't figure out why is that.

Hello  I want to know how do you visualize the training points randomly sampled in the geometry?

**Comment by gongsunlijiu** [comment]
> @lululxvi I made a simple test and I realized that the share of sampled points that goes to the complex non-homogeneous little sector is less than 10% of the points generated, i.e. the 90% of the training points are placed in the easy homogeneous big sector!
> 
> This is a pretty big waste, so I decided to anchor the points by myself with the `random_points` generator:
> 
> ![ic](https://user-images.githubusercontent.com/79721174/132367764-f7d24878-b954-4f41-b4ac-9e8ffecbc2fe.png)
> 
> With such training points' number and distribution - with the `[256]*5` NN I'm finally using - the full gradient descent makes totally sense and I'm sure that the learning of the IC will be perfect.
> 
> Still, the spatio-temporal problem of PDE and BC remains. The highly irregular distribution of IC forces me to choose the training points, I'll anchor the points by myself but I can't process a full gradient descent with that amount of points, so I'd ask you:
> 
> * say that I anchor a zillion points in the spatio-temporal domain, would it be possible for DeepXDE to split that array in small batches? I would like to:
>   
>   1. create the giant array of training point for IC, BC, Domain
>   2. ask DeepXDE to take 1,000 points, train 'em for 100 epochs, then take the next 1,000. So on till the end of the array.
> 
> Is it possible? Maybe every small-batch should include points from every component (PDE and BC and IC)?
> 
> Honestly I can't see another way to tacke this problem.

@Ryszard2 Hello, I see your question about the sampling points, which is very effective. 
I am now thinking about the same question and would like to ask you how you customize the sampling points to achieve different sampling densities for narrow and wide areas. Thank you very much
